akka-docs-dev/rst/scala/stream-parallelism.rst

```diff
@@ -0,0 +1,100 @@
+.. _stream-parallelism-scala:
+
+##########################
+Pipelining and Parallelism
+##########################
+
+Akka Streams processing stages (be it simple operators on Flows and Sources or graph junctions) are executed
+concurrently by default. This is realized by mapping each of the processing stages to a dedicated actor internally.
+
+We will illustrate through the example of pancake cooking how streams can be used for various processing patterns,
+exploiting the available parallelism on modern computers. The setting is the following: both Patrik and Roland
+likes to make pancakes, but they need to produce sufficient amount in a cooking session to make all of the children
+happy. To increase their pancake production throughput they use two frying pans. How they organize their pancake
+processing is markedly different.
+
+Pipelining
+----------
+
+Roland uses the two frying pans in an asymmetric fashion. The first pan is only used to fry one side of the
+pancake then the half-finished pancake is flipped into the second pan for the finishing fry on the other side.
+Once the first frying pan becomes available it gets a new scoop of batter. As an effect, most of the time there
+are two pancakes being cooked at the same time, one being cooked on its first side and the second being cooked to
+completion.
+This is how this setup would look like implemented as a stream:
+
+.. includecode:: code/docs/stream/FlowParallelismDocSpec.scala#pipelining
+
+The two ``map`` stages in sequence (encapsulated in the "frying pan" flows) will be executed in a pipelined way,
+basically doing the same as Roland with his frying pans:
+
+ 1. A ScoopOfBatter enters fryingPan1
+ 2. fryingPan1 emits a HalfCookedPancake once fryingPan2 becomes available
+ 3. fryingPan2 takes the HalfCookedPancake
+ 4. at this point fryingPan1 already takes the next scoop, without waiting for fryingPan2 to finish
+
+The benefit of pipelining is that it can be applied to any sequence of processing steps that are otherwise not
+parallelisable (for example because the result of a processing step depends on all the information from the previous
+step). One drawback is that if the processing times of the stages are very different then some of the stages will not
+be able to operate at full throughput because they will wait on a previous or subsequent stage most of the time. In the
+pancake example frying the second half of the pancake is usually faster than frying the first half, fryingPan2 will
+not be able to operate at full capacity.
```

Comment:
Awesome! How did you know that that is _exactly_ how I make pancakes? Let me also contribute the reason for this seemingly suboptimal procedure: I prefer the temperature of the second pan to be slightly lower than the first in order to achieve a more homogenous result.